examples/seq2seq/train.py

```diff
@@ -156,17 +155,19 @@ def apply(self,
       return x
 
     # No teacher forcing, feeding actual output back into the decoder.
-    output = np.zeros((max_output_len, batch_size, vocab_size))
-    for i in range(max_output_len):
-      decoder_inputs = np.expand_dims(decoder_inputs, axis=1)
-      carry, decoder_inputs = decoder(carry, decoder_inputs)
-      decoder_inputs = decoder_inputs.squeeze()
-      # Set next inputs to {0,1} to ensure they are the same as in training.
-      decoder_inputs = np.array([
-          decoder_inputs[j] == max(decoder_inputs[j]) for j in range(batch_size)
-      ])
-      output[:, i] = decoder_inputs
-    return output
+    next_inputs = jnp.zeros((batch_size, vocab_size))
+    output = []
+    for _ in range(max_output_len):
+      next_inputs = next_inputs[:, np.newaxis]
+      carry, decoder_outputs = decoder(carry, next_inputs)
+      decoder_outputs = decoder_outputs.squeeze()
+      output.append(decoder_outputs)
+      # Select the argmax as the next input.
+      next_inputs = jnp.equal(
+          decoder_outputs,
+          jnp.max(decoder_outputs, axis=-1)[:, None]
```

Comment:
Just for my understanding: axis=1 would also work here, right? What the intuition behind choosing -1?

Reply:
That's correct. I typically like to use axis=-1 in these cases since that is _almost always_ the logit/label dimension whereas sometimes dimension 1 is batch, sometimes it is sequence position, etc. Of course, we should try to be consistent with this.